### PR TITLE
gNOI Mock OS Generator library

### PR DESCRIPTION
**Pull request by Jared-Mullin**
Issue #99

### CODE REVIEW

File: utils/osgenerator/osgenerator.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	data.Write(buf)
+	hash := md5.Sum(data.Bytes())
+	data.Write(hash[:])
+	writer := bufio.NewWriter(file)
+	_, err = data.WriteTo(writer)
+	return err
+}
+
+// parseFilesize converts filesize to bytes.
+func parseFilesize(size int, unit rune) (int, error) {
```
Comment (loriscn12) [inline utils/osgenerator/osgenerator.go:52]: can this be made a little more descriptive? like: anyToBytes()

Comment (Jared-Mullin) [reply utils/osgenerator/osgenerator.go:52]: No problem, will do

File: utils/osgenerator/osgenerator.go [hunk 2]
```diff
@@ -36,16 +38,31 @@ func GenerateOS(filename, version string, unit rune, size int) error {
...
+	hash := md5.Sum(buf)
+	cookieBuf := make([]byte, 16)
+	rand.Read(cookieBuf)
+	mockOs := &pb.MockOS{
+		Version: version,
+		Cookie:  fmt.Sprintf("%x", cookieBuf),
+		Data:    buf,
+		Hash:    hash[:],
+	}
+	out, err := proto.Marshal(mockOs)
```
Comment (loriscn12) [inline utils/osgenerator/osgenerator.go]: optional (up to you): I would do it as follow if not reusing the var:

```suggestion
	out, err := proto.Marshal(&pb.MockOS{
		Version: version,
		Cookie:  fmt.Sprintf("%x", cookieBuf),
		Data:    buf,
		Hash:    hash[:],
	})
```

File: utils/osgenerator/osgenerator.go [hunk 3]
```diff
@@ -36,16 +38,31 @@ func GenerateOS(filename, version string, unit rune, size int) error {
...
+	if err != nil {
+		return err
+	}
 	writer := bufio.NewWriter(file)
-	_, err = data.WriteTo(writer)
-	return err
+	_, err = writer.Write(out)
+	if err != nil {
+		return err
+	}
```
Comment (loriscn12) [inline utils/osgenerator/osgenerator.go:61]: remove one line if you don't need the variable:

```suggestion
	if _, err = writer.Write(out); err != nil {
		return err
	}
```

File: utils/osgenerator/osgenerator.go [hunk 4]
```diff
@@ -36,16 +38,31 @@ func GenerateOS(filename, version string, unit rune, size int) error {
...
+	err = writer.Flush()
+	if err != nil {
+		return err
+	}
```
Comment (loriscn12) [inline utils/osgenerator/osgenerator.go:65]: same

File: utils/osgenerator/pb/mockos.proto [hunk 5]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+package pb;
+
```
Comment (loriscn12) [inline utils/osgenerator/pb/mockos.proto:4]: add message description

File: utils/osgenerator/pb/mockos.proto [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+message MockOS {
+    string version = 1;
+    string cookie = 2;
+    bytes data  = 3;
+    bytes hash = 4;
+    bool supported = 5;
```
Comment (samribeiro) [inline utils/osgenerator/pb/mockos.proto:10]: Add a comment stating that this flag is used to say that the OS image is valid but not supported by the Target.

File: gnoi_mockos/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# gNOI Mock OS Generator
+
+A simple shell binary generates a Mock OS package for use by a gNOI client.
```
Comment (samribeiro) [inline gnoi_mockos/README.md:3]: to use with the gNOI_target

File: gnoi_mockos/gnoi_mockos.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+	log "github.com/golang/glog"
+	"github.com/google/gnxi/utils/mockos"
+)
+
+var (
+	file      = flag.String("file", "", "The name and path of the OS file")
+	version   = flag.String("version", "1.0a", "The version of the OS package")
+	size      = flag.String("size", "", "The size of the OS package's data, e.g 10M")
+	supported = flag.Bool("supported", true, "Determines if the OS package is supported by the mock target")
```
Comment (samribeiro) [inline gnoi_mockos/gnoi_mockos.go:26]: unsupported

Comment (samribeiro) [inline gnoi_mockos/gnoi_mockos.go:26]: "I true, the OS package is valid but not supported by the Target"

File: utils/mockos/mockos_test.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+				},
+			},
+			want: false,
+		},
+	}
+
+	for _, test := range tests {
+		got := test.os.CheckHash()
+		if got != test.want {
+			t.Errorf("Expected %v, got %v", test.want, got)
```
Comment (loriscn12) [inline utils/mockos/mockos_test.go:43]: s/Expected/want

File: utils/mockos/mockos_test.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+			os: &OS{
+				MockOS: pb.MockOS{
+					Version:   "2.2b",
+					Cookie:    "yyhnueaolrasdfasdgdfsgdflkjsakfsdkoh",
+					Padding:   []byte("These are some bits to pad things out a lot!"),
+					Supported: false,
+					Hash:      []byte("Bad Hash"),
+				},
+			},
+			want: false,
```
Comment (loriscn12) [inline utils/mockos/mockos_test.go:36]: nit: you can omit false as that's by default, 1 line back :)

File: utils/mockos/mockos.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		return err
+	}
+	buf := make([]byte, bufferSize)
+	rand.Read(buf)
+	cookieBuf := make([]byte, 16)
+	rand.Read(cookieBuf)
+	mockOs := &OS{MockOS: pb.MockOS{
+		Version:   version,
+		Cookie:    fmt.Sprintf("%x", cookieBuf),
```
Comment (samribeiro) [inline utils/mockos/mockos.go:64]: The cookie is best if it is a const in this file.

File: gnoi_mockos/gnoi_mockos.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+	log "github.com/golang/glog"
+	"github.com/google/gnxi/utils/mockos"
+)
+
+var (
+	file      = flag.String("file", "", "The name and path of the OS file")
+	version   = flag.String("version", "", "The version of the OS package")
+	size      = flag.String("size", "1M", "The size of the OS package's data, e.g 10M")
+	supported = flag.Bool("unsupported", false, "If true, the os package is valid but unsupported by the target")
```
Comment (samribeiro) [inline gnoi_mockos/gnoi_mockos.go:27]: worth naming it "unsupported" everywhere, including the proto =)

File: utils/mockos/mockos.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	cookie = "cookiestring"
+)
+
+type OS struct {
+	pb.MockOS
+}
+
+// Hash calculates the hash of the MockOS and embeds it in the package.
+func (os *OS) Hash() {
+	temp := calcHash(os)
```
Comment (samribeiro) [inline utils/mockos/mockos.go:37]: one line functions are ok =)

	os.MockOS.Hash = calcHash(os)

File: utils/mockos/mockos.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	os.MockOS.Hash = temp[:]
+}
+
+// CheckHash calculates the hash of the MockOS and checks against the embedded hash.
+func (os *OS) CheckHash() bool {
+	temp := calcHash(os)
```
Comment (samribeiro) [inline utils/mockos/mockos.go:43]: return bytes.Compare(os.MockOS.Hash, calcHash(os)) == 0

File: gnoi_mockos/gnoi_mockos.go (hunk 12 above)
Comment (Jared-Mullin) [reply gnoi_mockos/gnoi_mockos.go]: I think I caught them all this time, sorry about that

File: utils/mockos/mockos.go [hunk 15]
```diff
@@ -103,6 +103,5 @@ func calcHash(os *OS) []byte {
 	bb = append(bb, []byte(os.MockOS.Cookie)...)
 	bb = append(bb, []byte(os.MockOS.Padding)...)
 	bb = append(bb, map[bool]byte{false: 0, true: 1}[os.MockOS.Unsupported])
-	hash := md5.Sum(bb)
-	return hash[:]
+	return md5.Sum(bb)
```
Comment (Jared-Mullin) [inline utils/mockos/mockos.go:106]: md5.Sum() returns a [16]byte so I think it's causing an error

Comment (samribeiro) [reply utils/mockos/mockos.go:106]: indeed

**Pull request merged**